tensorboard/uploader/uploader.py

```diff
@@ -309,26 +271,98 @@ def _run_values(self, run_to_events):
+    """Helper class for building requests that fit under a size limit.
+
+    This class accumulates a current request.  `add_event(...)` may or may not
+    send the request (and start a new one).  After all `add_event(...)` calls
+    are complete, a final call to `flush()` is needed to send the final request.
 
-        Returns:
-          The active request, or `None` if after pruning the request
-          contains no data.
+    This class is not threadsafe. Use external synchronization if calling its
+    methods concurrently.
+    """
+
+    def __init__(self, experiment_id, api, rpc_rate_limiter):
+        if experiment_id is None:
+            raise ValueError("experiment_id cannot be None")
+        self._experiment_id = experiment_id
+        self._api = api
+        self._rpc_rate_limiter = rpc_rate_limiter
+        # A lower bound on the number of bytes that we may yet add to the
+        # request.
+        self._byte_budget = None  # type: int
+
+        self._runs = {}  # cache: map from run name to `Run` proto in request
+        self._tags = (
+            {}
+        )  # cache: map from `(run, tag)` to `Tag` proto in run in request
+        self._new_request()
+
+    def _new_request(self):
+        """Allocates a new request and refreshes the budget."""
+        self._request = write_service_pb2.WriteScalarRequest()
+        self._runs.clear()
+        self._tags.clear()
+        self._byte_budget = _MAX_REQUEST_LENGTH_BYTES
+        self._request.experiment_id = self._experiment_id
+        self._byte_budget -= self._request.ByteSize()
+        if self._byte_budget < 0:
+            raise RuntimeError("Byte budget too small for experiment ID")
+
+    def add_event(self, run_name, event, value, metadata, is_retry=False):
+        """Attempts to add the given event to the current request.
+
+        If the event cannot be added to the current request because the byte
+        budget is exhausted, the request is flushed, and the event is added
+        to the next request.
+        """
+        try:
+            run_proto = self._runs.get(run_name)
+            if run_proto is None:
+                run_proto = self._create_run(run_name)
+                self._runs[run_name] = run_proto
+            tag_proto = self._tags.get((run_name, value.tag))
+            if tag_proto is None:
+                tag_proto = self._create_tag(run_proto, value.tag, metadata)
+                self._tags[(run_name, value.tag)] = tag_proto
+            self._create_point(tag_proto, event, value)
+        except _OutOfSpaceError:
+            if is_retry:
+                raise RuntimeError("add_event failed despite flush")
+            self.flush()
+            # Try again.  This attempt should never produce OutOfSpaceError
+            # because we just flushed.  Nonetheless we use the is_retry
+            # mechanism to enforce that we don't recurse more than once.
+            self.add_event(run_name, event, value, metadata, is_retry=True)
```

Comment:
Interesting approach. I will add this to my toolbox of “ways to deal
with pagination problems”. :-)

Reply:
FWIW I don't think we really need recursion here (and the use of it leaks into the "public" API of this class, though I recognize the class itself is still an implementation detail).

We can just factor out the body of the `try` into a private helper, and then do

```python
try:
  self._internal_add_event(run_name, event, value, metadata)
except _OutOfSpaceError:
  self.flush()
   # Try again.  This attempt should never produce OutOfSpaceError because we just flushed.
  self._internal_add_event(run_name, event, value, metadata)
```

Reply:
Good point, done, with the modification that we have to catch `_OutOfSpaceError` on the retry too even though it's an edge case.  E.g. we have a test for the resulting `RuntimeError` when the run name alone is larger than the byte budget.